Windows stack to heap crash since LLVM update

This was a corral nightly breakage check on Windows.

2026-03-06T00:41:41.2298337Z Current runner version: '2.332.0'
2026-03-06T00:41:41.2331331Z ##[group]Runner Image Provisioner
2026-03-06T00:41:41.2332426Z Hosted Compute Agent
2026-03-06T00:41:41.2333315Z Version: 20260213.493
2026-03-06T00:41:41.2334272Z Commit: 5c115507f6dd24b8de37d8bbe0bb4509d0cc0fa3
2026-03-06T00:41:41.2335301Z Build Date: 2026-02-13T00:28:41Z
2026-03-06T00:41:41.2336360Z Worker ID: {e426925a-25dc-4209-82b4-842712061ce5}
2026-03-06T00:41:41.2337783Z Azure Region: northcentralus
2026-03-06T00:41:41.2338725Z ##[endgroup]
2026-03-06T00:41:41.2340731Z ##[group]Operating System
2026-03-06T00:41:41.2341675Z Microsoft Windows Server 2025
2026-03-06T00:41:41.2342585Z 10.0.26100
2026-03-06T00:41:41.2343276Z Datacenter
2026-03-06T00:41:41.2344032Z ##[endgroup]
2026-03-06T00:41:41.2344804Z ##[group]Runner Image
2026-03-06T00:41:41.2345612Z Image: windows-2025
2026-03-06T00:41:41.2346430Z Version: 20260302.43.1
2026-03-06T00:41:41.2348731Z Included Software: https://github.com/actions/runner-images/blob/win25/20260302.43/images/windows/Windows2025-Readme.md
2026-03-06T00:41:41.2351373Z Image Release: https://github.com/actions/runner-images/releases/tag/win25%2F20260302.43
2026-03-06T00:41:41.2352998Z ##[endgroup]
2026-03-06T00:41:41.2354673Z ##[group]GITHUB_TOKEN Permissions
2026-03-06T00:41:41.2357098Z Metadata: read
2026-03-06T00:41:41.2358339Z Packages: read
2026-03-06T00:41:41.2359088Z ##[endgroup]
2026-03-06T00:41:41.2361791Z Secret source: Actions
2026-03-06T00:41:41.2362938Z Prepare workflow directory
2026-03-06T00:41:41.2762769Z Prepare all required actions
2026-03-06T00:41:41.2800821Z Getting action download info
2026-03-06T00:41:41.5314991Z Download action repository 'actions/checkout@v4.1.1' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
2026-03-06T00:41:41.6589750Z Download action repository 'zulip/github-actions-zulip@e4c8f27c732ba9bd98ac6be0583096dea82feea5' (SHA:e4c8f27c732ba9bd98ac6be0583096dea82feea5)
2026-03-06T00:41:42.4888905Z Complete job name: Verify main against the latest ponyc on Windows
2026-03-06T00:41:42.6151304Z ##[group]Run actions/checkout@v4.1.1
2026-03-06T00:41:42.6152363Z with:
2026-03-06T00:41:42.6152776Z   repository: ponylang/corral
2026-03-06T00:41:42.6153446Z   token: ***
2026-03-06T00:41:42.6153829Z   ssh-strict: true
2026-03-06T00:41:42.6154257Z   persist-credentials: true
2026-03-06T00:41:42.6154717Z   clean: true
2026-03-06T00:41:42.6155118Z   sparse-checkout-cone-mode: true
2026-03-06T00:41:42.6155606Z   fetch-depth: 1
2026-03-06T00:41:42.6156012Z   fetch-tags: false
2026-03-06T00:41:42.6156417Z   show-progress: true
2026-03-06T00:41:42.6156825Z   lfs: false
2026-03-06T00:41:42.6157196Z   submodules: false
2026-03-06T00:41:42.6157615Z   set-safe-directory: true
2026-03-06T00:41:42.6158297Z ##[endgroup]
2026-03-06T00:41:42.7531999Z Syncing repository: ponylang/corral
2026-03-06T00:41:42.7533785Z ##[group]Getting Git version info
2026-03-06T00:41:42.7534409Z Working directory is 'D:\a\corral\corral'
2026-03-06T00:41:42.8451588Z [command]"C:\Program Files\Git\bin\git.exe" version
2026-03-06T00:41:43.0512820Z git version 2.53.0.windows.1
2026-03-06T00:41:43.0587422Z ##[endgroup]
2026-03-06T00:41:43.0613291Z Temporarily overriding HOME='D:\a\_temp\d071f7c2-a49b-45aa-baab-d3f2fce23675' before making global git config changes
2026-03-06T00:41:43.0616770Z Adding repository directory to the temporary git global config as a safe directory
2026-03-06T00:41:43.0623938Z [command]"C:\Program Files\Git\bin\git.exe" config --global --add safe.directory D:\a\corral\corral
2026-03-06T00:41:43.0953509Z Deleting the contents of 'D:\a\corral\corral'
2026-03-06T00:41:43.0961156Z ##[group]Initializing the repository
2026-03-06T00:41:43.0969280Z [command]"C:\Program Files\Git\bin\git.exe" init D:\a\corral\corral
2026-03-06T00:41:43.1459221Z Initialized empty Git repository in D:/a/corral/corral/.git/
2026-03-06T00:41:43.1511186Z [command]"C:\Program Files\Git\bin\git.exe" remote add origin https://github.com/ponylang/corral
2026-03-06T00:41:43.1932271Z ##[endgroup]
2026-03-06T00:41:43.1935425Z ##[group]Disabling automatic garbage collection
2026-03-06T00:41:43.1936198Z [command]"C:\Program Files\Git\bin\git.exe" config --local gc.auto 0
2026-03-06T00:41:43.2317378Z ##[endgroup]
2026-03-06T00:41:43.2318066Z ##[group]Setting up auth
2026-03-06T00:41:43.2334981Z [command]"C:\Program Files\Git\bin\git.exe" config --local --name-only --get-regexp core\.sshCommand
2026-03-06T00:41:43.2787209Z [command]"C:\Program Files\Git\bin\git.exe" submodule foreach --recursive "sh -c \"git config --local --name-only --get-regexp 'core\.sshCommand' && git config --local --unset-all 'core.sshCommand' || :\""
2026-03-06T00:41:44.2986414Z [command]"C:\Program Files\Git\bin\git.exe" config --local --name-only --get-regexp http\.https\:\/\/github\.com\/\.extraheader
2026-03-06T00:41:44.3333048Z [command]"C:\Program Files\Git\bin\git.exe" submodule foreach --recursive "sh -c \"git config --local --name-only --get-regexp 'http\.https\:\/\/github\.com\/\.extraheader' && git config --local --unset-all 'http.https://github.com/.extraheader' || :\""
2026-03-06T00:41:44.9075688Z [command]"C:\Program Files\Git\bin\git.exe" config --local http.https://github.com/.extraheader "AUTHORIZATION: basic ***"
2026-03-06T00:41:44.9413596Z ##[endgroup]
2026-03-06T00:41:44.9414404Z ##[group]Fetching the repository
2026-03-06T00:41:44.9426890Z [command]"C:\Program Files\Git\bin\git.exe" -c protocol.version=2 fetch --no-tags --prune --no-recurse-submodules --depth=1 origin +1c22097579d4b52b774f54fcc73f87da6b075e2d:refs/remotes/origin/main
2026-03-06T00:41:45.6037465Z From https://github.com/ponylang/corral
2026-03-06T00:41:45.6038607Z  * [new ref]         1c22097579d4b52b774f54fcc73f87da6b075e2d -> origin/main
2026-03-06T00:41:45.6292536Z ##[endgroup]
2026-03-06T00:41:45.6293230Z ##[group]Determining the checkout info
2026-03-06T00:41:45.6295671Z ##[endgroup]
2026-03-06T00:41:45.6296783Z ##[group]Checking out the ref
2026-03-06T00:41:45.6308764Z [command]"C:\Program Files\Git\bin\git.exe" checkout --progress --force -B main refs/remotes/origin/main
2026-03-06T00:41:45.7265706Z Switched to a new branch 'main'
2026-03-06T00:41:45.7289499Z branch 'main' set up to track 'origin/main'.
2026-03-06T00:41:45.7345044Z ##[endgroup]
2026-03-06T00:41:45.7765709Z [command]"C:\Program Files\Git\bin\git.exe" log -1 --format='%H'
2026-03-06T00:41:45.8066788Z '1c22097579d4b52b774f54fcc73f87da6b075e2d'
2026-03-06T00:41:45.8583811Z ##[group]Run Invoke-WebRequest https://dl.cloudsmith.io/public/ponylang/nightlies/raw/versions/latest/ponyc-x86-64-pc-windows-msvc.zip -OutFile C:\ponyc.zip;
2026-03-06T00:41:45.8586106Z [36;1mInvoke-WebRequest https://dl.cloudsmith.io/public/ponylang/nightlies/raw/versions/latest/ponyc-x86-64-pc-windows-msvc.zip -OutFile C:\ponyc.zip;[0m
2026-03-06T00:41:45.8587562Z [36;1mExpand-Archive -Path C:\ponyc.zip -DestinationPath C:\ponyc;[0m
2026-03-06T00:41:45.8588449Z [36;1m$env:PATH = 'C:\ponyc\bin;' + $env:PATH;[0m
2026-03-06T00:41:45.8588987Z [36;1m.\make.ps1 -Command unit-tests 2>&1[0m
2026-03-06T00:41:45.8792579Z shell: C:\Program Files\PowerShell\7\pwsh.EXE -command ". '{0}'"
2026-03-06T00:41:45.8793447Z ##[endgroup]
2026-03-06T00:42:02.0990623Z fatal: Needed a single revision
2026-03-06T00:42:02.1157708Z Configuration:    Release
2026-03-06T00:42:02.1161664Z Version:          0.9.2-
2026-03-06T00:42:02.1164036Z Root directory:   D:\a\corral\corral
2026-03-06T00:42:02.1167113Z Source directory: D:\a\corral\corral\corral
2026-03-06T00:42:02.1170167Z Build directory:  D:\a\corral\corral\build\release
2026-03-06T00:42:02.1552332Z D:\a\corral\corral\corral\version.pony.in -> D:\a\corral\corral\corral\version.pony
2026-03-06T00:42:02.2521920Z Building builtin -> C:\ponyc\packages\builtin
2026-03-06T00:42:02.3148808Z Building  -> C:\ponyc\bin
2026-03-06T00:42:02.3184678Z Building builtin -> C:\ponyc\packages\builtin
2026-03-06T00:42:02.3780855Z Building D:\a\corral\corral\corral -> D:\a\corral\corral\corral
2026-03-06T00:42:02.3887344Z Building cli -> C:\ponyc\packages\cli
2026-03-06T00:42:02.4124750Z Building collections -> C:\ponyc\packages\collections
2026-03-06T00:42:02.4615463Z Building pony_test -> C:\ponyc\packages\pony_test
2026-03-06T00:42:02.4722634Z Building time -> C:\ponyc\packages\time
2026-03-06T00:42:02.4852559Z Building random -> C:\ponyc\packages\random
2026-03-06T00:42:02.5215872Z Building collections/persistent -> C:\ponyc\packages\collections\persistent
2026-03-06T00:42:02.5571331Z Building buffered -> C:\ponyc\packages\buffered
2026-03-06T00:42:02.5770985Z Building cmd -> D:\a\corral\corral\corral\cmd
2026-03-06T00:42:02.5915443Z Building ../logger -> D:\a\corral\corral\corral\logger
2026-03-06T00:42:02.5949509Z Building promises -> C:\ponyc\packages\promises
2026-03-06T00:42:02.6036834Z Building process -> C:\ponyc\packages\process
2026-03-06T00:42:02.6294640Z Building backpressure -> C:\ponyc\packages\backpressure
2026-03-06T00:42:02.6307978Z Building files -> C:\ponyc\packages\files
2026-03-06T00:42:02.6745805Z Building term -> C:\ponyc\packages\term
2026-03-06T00:42:02.6865899Z Building strings -> C:\ponyc\packages\strings
2026-03-06T00:42:02.6895763Z Building signals -> C:\ponyc\packages\signals
2026-03-06T00:42:02.6987298Z Building capsicum -> C:\ponyc\packages\capsicum
2026-03-06T00:42:02.7083371Z Building itertools -> C:\ponyc\packages\itertools
2026-03-06T00:42:02.7413381Z Building ../bundle -> D:\a\corral\corral\corral\bundle
2026-03-06T00:42:02.7533746Z Building ../json -> D:\a\corral\corral\corral\json
2026-03-06T00:42:02.7712403Z Building format -> C:\ponyc\packages\format
2026-03-06T00:42:02.7874914Z Building ../semver/utils -> D:\a\corral\corral\corral\semver\utils
2026-03-06T00:42:02.7915708Z Building ../util -> D:\a\corral\corral\corral\util
2026-03-06T00:42:02.7963501Z Building ../vcs -> D:\a\corral\corral\corral\vcs
2026-03-06T00:42:02.8018589Z Building debug -> C:\ponyc\packages\debug
2026-03-06T00:42:02.8062999Z Building ../semver/range -> D:\a\corral\corral\corral\semver\range
2026-03-06T00:42:02.8084298Z Building ../version -> D:\a\corral\corral\corral\semver\version
2026-03-06T00:42:02.8139653Z Building ../semver/solver -> D:\a\corral\corral\corral\semver\solver
2026-03-06T00:42:02.8234751Z Building ../archive -> D:\a\corral\corral\corral\archive
2026-03-06T00:42:02.8257583Z Building ../mort -> D:\a\corral\corral\corral\mort
2026-03-06T00:42:07.3410594Z Generating
2026-03-06T00:42:07.3467726Z  Reachability
2026-03-06T00:42:08.2979588Z  Selector painting
2026-03-06T00:42:08.3026895Z  Data prototypes
2026-03-06T00:42:08.3107395Z  Data types
2026-03-06T00:42:08.3113497Z  Function prototypes
2026-03-06T00:42:08.3371668Z  Functions
2026-03-06T00:42:08.7265692Z  Descriptors
2026-03-06T00:42:14.9009727Z Verifying
2026-03-06T00:42:14.9569798Z Writing D:\a\corral\corral\build\release\corral.obj
2026-03-06T00:42:21.3891384Z Linking D:\a\corral\corral\build\release\corral.exe
2026-03-06T00:42:22.3569275Z    Creating library D:\a\corral\corral\build\release\corral.lib and object D:\a\corral\corral\build\release\corral.exp
2026-03-06T00:42:23.7114792Z Error:
2026-03-06T00:42:23.7115248Z no source files in package ''
2026-03-06T00:42:23.8169746Z Building builtin -> C:\ponyc\packages\builtin
2026-03-06T00:42:23.8773408Z Building  -> C:\ponyc\bin
2026-03-06T00:42:23.8810445Z Building builtin -> C:\ponyc\packages\builtin
2026-03-06T00:42:23.9372770Z Building D:\a\corral\corral\corral\test -> D:\a\corral\corral\corral\test
2026-03-06T00:42:23.9482040Z Building cli -> C:\ponyc\packages\cli
2026-03-06T00:42:23.9730325Z Building collections -> C:\ponyc\packages\collections
2026-03-06T00:42:24.0185914Z Building pony_test -> C:\ponyc\packages\pony_test
2026-03-06T00:42:24.0298649Z Building time -> C:\ponyc\packages\time
2026-03-06T00:42:24.0444292Z Building random -> C:\ponyc\packages\random
2026-03-06T00:42:24.0840305Z Building collections/persistent -> C:\ponyc\packages\collections\persistent
2026-03-06T00:42:24.1220578Z Building buffered -> C:\ponyc\packages\buffered
2026-03-06T00:42:24.1424671Z Building files -> C:\ponyc\packages\files
2026-03-06T00:42:24.1885939Z Building term -> C:\ponyc\packages\term
2026-03-06T00:42:24.1982527Z Building promises -> C:\ponyc\packages\promises
2026-03-06T00:42:24.2078875Z Building strings -> C:\ponyc\packages\strings
2026-03-06T00:42:24.2108156Z Building signals -> C:\ponyc\packages\signals
2026-03-06T00:42:24.2189399Z Building capsicum -> C:\ponyc\packages\capsicum
2026-03-06T00:42:24.2256186Z Building ../util -> D:\a\corral\corral\corral\util
2026-03-06T00:42:24.2290674Z Building ../logger -> D:\a\corral\corral\corral\logger
2026-03-06T00:42:24.2333084Z Building backpressure -> C:\ponyc\packages\backpressure
2026-03-06T00:42:24.2346141Z Building process -> C:\ponyc\packages\process
2026-03-06T00:42:24.2605374Z Building itertools -> C:\ponyc\packages\itertools
2026-03-06T00:42:24.2839529Z Building ../vcs -> D:\a\corral\corral\corral\vcs
2026-03-06T00:42:24.2890093Z Building debug -> C:\ponyc\packages\debug
2026-03-06T00:42:24.2939686Z Building ../bundle -> D:\a\corral\corral\corral\bundle
2026-03-06T00:42:24.3019080Z Building ../json -> D:\a\corral\corral\corral\json
2026-03-06T00:42:24.3204194Z Building format -> C:\ponyc\packages\format
2026-03-06T00:42:24.3382166Z Building ../semver/utils -> D:\a\corral\corral\corral\semver\utils
2026-03-06T00:42:24.3432732Z Building integration -> D:\a\corral\corral\corral\test\integration
2026-03-06T00:42:24.3529313Z Building ../.. -> D:\a\corral\corral\corral
2026-03-06T00:42:24.3556325Z Building cmd -> D:\a\corral\corral\corral\cmd
2026-03-06T00:42:24.3719019Z Building ../semver/range -> D:\a\corral\corral\corral\semver\range
2026-03-06T00:42:24.3741983Z Building ../version -> D:\a\corral\corral\corral\semver\version
2026-03-06T00:42:24.3806911Z Building ../semver/solver -> D:\a\corral\corral\corral\semver\solver
2026-03-06T00:42:24.3898803Z Building ../archive -> D:\a\corral\corral\corral\archive
2026-03-06T00:42:24.3920919Z Building ../mort -> D:\a\corral\corral\corral\mort
2026-03-06T00:42:28.3032879Z Generating
2026-03-06T00:42:28.3081701Z  Reachability
2026-03-06T00:42:29.0792068Z  Selector painting
2026-03-06T00:42:29.0834062Z  Data prototypes
2026-03-06T00:42:29.0905239Z  Data types
2026-03-06T00:42:29.0911568Z  Function prototypes
2026-03-06T00:42:29.1158487Z  Functions
2026-03-06T00:42:29.4390643Z  Descriptors
2026-03-06T00:42:33.5315694Z PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
2026-03-06T00:42:33.5316613Z Stack dump:
2026-03-06T00:42:33.5335892Z 0.	Running pass "require<GlobalsAA>,function(invalidate<AAManager>),require<ProfileSummaryAnalysis>,cgscc(devirt<4>(InlinerPass,PostOrderFunctionAttrsPass<skip-non-recursive-function-attrs>,ArgumentPromotionPass,OpenMPOptCGSCCPass,function<eager-inv;no-rerun>(SROAPass<modify-cfg>,EarlyCSEPass<memssa>,SpeculativeExecutionPass<only-if-divergent-target>,JumpThreadingPass,CorrelatedValuePropagationPass,SimplifyCFGPass<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,InstCombinePass<max-iterations=1;no-verify-fixpoint>,AggressiveInstCombinePass,LibCallsShrinkWrapPass,HeapToStack,TailCallElimPass,SimplifyCFGPass<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,ReassociatePass,ConstraintEliminationPass,loop-mssa(LoopInstSimplifyPass,LoopSimplifyCFGPass,LICMPass<no-allowspeculation>,LoopRotatePass<header-duplication;no-prepare-for-lto>,LICMPass<allowspeculation>,SimpleLoopUnswitchPass<nontrivial;trivial>),SimplifyCFGPass<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,InstCombinePass<max-iterations=1;no-verify-fixpoint>,loop(LoopIdiomRecognizePass,IndVarSimplifyPass,ExtraLoopPassManager<struct llvm::ShouldRunExtraSimpleLoopUnswitch>,LoopDeletionPass,LoopFullUnrollPass),SROAPass<modify-cfg>,VectorCombinePass,MergedLoadStoreMotionPass<no-split-footer-bb>,GVNPass<>,SCCPPass,BDCEPass,InstCombinePass<max-iterations=1;no-verify-fixpoint>,HeapToStack,JumpThreadingPass,CorrelatedValuePropagationPass,ADCEPass,MemCpyOptPass,DSEPass,MoveAutoInitPass,loop-mssa(LICMPass<allowspeculation>),CoroElidePass,DispatchPonyCtx,MergeMessageSend,SimplifyCFGPass<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,InstCombinePass<max-iterations=1;no-verify-fixpoint>,HeapToStack),PostOrderFunctionAttrsPass,function(require<ShouldNotRunFunctionPassesAnalysis>),CoroSplitPass,CoroAnnotationElidePass)),function(invalidate<ShouldNotRunFunctionPassesAnalysis>),cgscc(devirt<4>())" on module "test"
2026-03-06T00:42:33.5358004Z 1.	Running pass "cgscc(devirt<4>(InlinerPass,PostOrderFunctionAttrsPass<skip-non-recursive-function-attrs>,ArgumentPromotionPass,OpenMPOptCGSCCPass,function<eager-inv;no-rerun>(SROAPass<modify-cfg>,EarlyCSEPass<memssa>,SpeculativeExecutionPass<only-if-divergent-target>,JumpThreadingPass,CorrelatedValuePropagationPass,SimplifyCFGPass<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,InstCombinePass<max-iterations=1;no-verify-fixpoint>,AggressiveInstCombinePass,LibCallsShrinkWrapPass,HeapToStack,TailCallElimPass,SimplifyCFGPass<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,ReassociatePass,ConstraintEliminationPass,loop-mssa(LoopInstSimplifyPass,LoopSimplifyCFGPass,LICMPass<no-allowspeculation>,LoopRotatePass<header-duplication;no-prepare-for-lto>,LICMPass<allowspeculation>,SimpleLoopUnswitchPass<nontrivial;trivial>),SimplifyCFGPass<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;no-hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;no-sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,InstCombinePass<max-iterations=1;no-verify-fixpoint>,loop(LoopIdiomRecognizePass,IndVarSimplifyPass,ExtraLoopPassManager<struct llvm::ShouldRunExtraSimpleLoopUnswitch>,LoopDeletionPass,LoopFullUnrollPass),SROAPass<modify-cfg>,VectorCombinePass,MergedLoadStoreMotionPass<no-split-footer-bb>,GVNPass<>,SCCPPass,BDCEPass,InstCombinePass<max-iterations=1;no-verify-fixpoint>,HeapToStack,JumpThreadingPass,CorrelatedValuePropagationPass,ADCEPass,MemCpyOptPass,DSEPass,MoveAutoInitPass,loop-mssa(LICMPass<allowspeculation>),CoroElidePass,DispatchPonyCtx,MergeMessageSend,SimplifyCFGPass<bonus-inst-threshold=1;no-forward-switch-cond;switch-range-to-icmp;no-switch-to-lookup;keep-loops;hoist-common-insts;no-hoist-loads-stores-with-cond-faulting;sink-common-insts;speculate-blocks;simplify-cond-branch;no-speculate-unpredictables>,InstCombinePass<max-iterations=1;no-verify-fixpoint>,HeapToStack),PostOrderFunctionAttrsPass,function(require<ShouldNotRunFunctionPassesAnalysis>),CoroSplitPass,CoroAnnotationElidePass))" on module "test"
2026-03-06T00:42:33.5368689Z 2.	Running pass "HeapToStack" on function "TestGitParseTags_ref_apply_oo"
2026-03-06T00:42:33.5369054Z Exception Code: 0xC0000005
2026-03-06T00:42:33.6500811Z  #0 0x00007ff66fa9e0d4 pony_actor_stats (C:\ponyc\bin\ponyc.exe+0x14e0d4)
2026-03-06T00:42:33.6502140Z  #1 0x00007ff66fa9da6e pony_actor_stats (C:\ponyc\bin\ponyc.exe+0x14da6e)
2026-03-06T00:42:33.6504451Z  #2 0x00007ff66fb80422 pony_asio_event_set_writeable (C:\ponyc\bin\ponyc.exe+0x230422)
2026-03-06T00:42:33.6505333Z  #3 0x00007ff670201482 pony_asio_event_set_writeable (C:\ponyc\bin\ponyc.exe+0x8b1482)
2026-03-06T00:42:33.6506124Z  #4 0x00007ff6714f951d pony_asio_event_set_writeable (C:\ponyc\bin\ponyc.exe+0x1ba951d)
2026-03-06T00:42:33.6506926Z  #5 0x00007ff6702021d6 pony_asio_event_set_writeable (C:\ponyc\bin\ponyc.exe+0x8b21d6)
2026-03-06T00:42:33.6507717Z  #6 0x00007ff6714f9005 pony_asio_event_set_writeable (C:\ponyc\bin\ponyc.exe+0x1ba9005)
2026-03-06T00:42:33.6508483Z  #7 0x00007ff67066bd56 pony_asio_event_set_writeable (C:\ponyc\bin\ponyc.exe+0xd1bd56)
2026-03-06T00:42:33.6509265Z  #8 0x00007ff6714f9a38 pony_asio_event_set_writeable (C:\ponyc\bin\ponyc.exe+0x1ba9a38)
2026-03-06T00:42:33.6510095Z  #9 0x00007ff67068ae56 pony_asio_event_set_writeable (C:\ponyc\bin\ponyc.exe+0xd3ae56)
2026-03-06T00:42:33.6510871Z #10 0x00007ff6714fab25 pony_asio_event_set_writeable (C:\ponyc\bin\ponyc.exe+0x1baab25)
2026-03-06T00:42:33.6511656Z #11 0x00007ff670202162 pony_asio_event_set_writeable (C:\ponyc\bin\ponyc.exe+0x8b2162)
2026-03-06T00:42:33.6512421Z #12 0x00007ff66fb80682 pony_asio_event_set_writeable (C:\ponyc\bin\ponyc.exe+0x230682)
2026-03-06T00:42:33.6513552Z #13 0x00007ff670e3ccd0 pony_asio_event_set_writeable (C:\ponyc\bin\ponyc.exe+0x14eccd0)
2026-03-06T00:42:33.6514345Z #14 0x00007ff67066ba62 pony_asio_event_set_writeable (C:\ponyc\bin\ponyc.exe+0xd1ba62)
2026-03-06T00:42:33.6515100Z #15 0x00007ff66fb80682 pony_asio_event_set_writeable (C:\ponyc\bin\ponyc.exe+0x230682)
2026-03-06T00:42:33.6515819Z #16 0x00007ff66fa9cfce pony_actor_stats (C:\ponyc\bin\ponyc.exe+0x14cfce)
2026-03-06T00:42:33.6516491Z #17 0x00007ff66fa9ed36 pony_actor_stats (C:\ponyc\bin\ponyc.exe+0x14ed36)
2026-03-06T00:42:33.6517060Z #18 0x00007ff66fa8d143 (C:\ponyc\bin\ponyc.exe+0x13d143)
2026-03-06T00:42:33.6517528Z #19 0x00007ff66fa4a155 (C:\ponyc\bin\ponyc.exe+0xfa155)
2026-03-06T00:42:33.6517976Z #20 0x00007ff66f9f251f (C:\ponyc\bin\ponyc.exe+0xa251f)
2026-03-06T00:42:33.6518585Z #21 0x00007ff671911960 pony_asio_event_set_writeable (C:\ponyc\bin\ponyc.exe+0x1fc1960)
2026-03-06T00:42:33.6519272Z #22 0x00007ffc3451e8d7 (C:\Windows\System32\KERNEL32.DLL+0x2e8d7)
2026-03-06T00:42:33.6519839Z #23 0x00007ffc353cc53c (C:\Windows\SYSTEM32\ntdll.dll+0x8c53c)
2026-03-06T00:42:34.3247902Z [31;1mmake.ps1: [0mD:\a\_temp\b0361f08-5c21-44f9-964f-1348582c1bf1.ps1:5[0m
2026-03-06T00:42:34.3248795Z [31;1m[0m[36;1mLine |[0m
2026-03-06T00:42:34.3249786Z [31;1m[0m[36;1m[36;1m   5 | [0m [36;1m.\make.ps1 -Command unit-tests 2>&1[0m
2026-03-06T00:42:34.3250620Z [31;1m[0m[36;1m[36;1m[0m[36;1m[0m[36;1m     | [31;1m ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[0m
2026-03-06T00:42:34.3251482Z [31;1m[0m[36;1m[36;1m[0m[36;1m[0m[36;1m[31;1m[31;1m[36;1m     | [31;1mThe term 'D:\a\corral\corral\build\release\test.exe' is not recognized as a name of a cmdlet, function, script[0m
2026-03-06T00:42:34.3252706Z [31;1m[0m[36;1m[36;1m[0m[36;1m[0m[36;1m[31;1m[31;1m[36;1m[31;1m[36;1m     | [31;1mfile, or executable program. Check the spelling of the name, or if a path was included, verify that the path is[0m
2026-03-06T00:42:34.3253647Z [31;1m[0m[36;1m[36;1m[0m[36;1m[0m[36;1m[31;1m[31;1m[36;1m[31;1m[36;1m[31;1m[36;1m     | [31;1mcorrect and try again.[0m
2026-03-06T00:42:34.3701197Z ##[error]Process completed with exit code 1.
2026-03-06T00:42:34.3954450Z ##[group]Run zulip/github-actions-zulip/send-message@e4c8f27c732ba9bd98ac6be0583096dea82feea5
2026-03-06T00:42:34.3954960Z with:
2026-03-06T00:42:34.3955376Z   api-key: ***
2026-03-06T00:42:34.3955643Z   email: ***
2026-03-06T00:42:34.3955887Z   organization-url: https://ponylang.zulipchat.com/
2026-03-06T00:42:34.3956183Z   to: notifications
2026-03-06T00:42:34.3956339Z   type: stream
2026-03-06T00:42:34.3956524Z   topic: ponylang/corral scheduled job failure
2026-03-06T00:42:34.3956903Z   content: https://github.com/ponylang/corral/actions/runs/22743479385 failed.
2026-03-06T00:42:34.3957772Z ##[endgroup]
2026-03-06T00:42:34.8404050Z Message successfully sent with id: 577625363
2026-03-06T00:42:34.8760130Z Post job cleanup.
2026-03-06T00:42:35.0834271Z [command]"C:\Program Files\Git\bin\git.exe" version
2026-03-06T00:42:35.1217517Z git version 2.53.0.windows.1
2026-03-06T00:42:35.1310035Z Temporarily overriding HOME='D:\a\_temp\9bb5d74c-a0f9-4f3e-b5aa-f585380d8c73' before making global git config changes
2026-03-06T00:42:35.1311270Z Adding repository directory to the temporary git global config as a safe directory
2026-03-06T00:42:35.1318747Z [command]"C:\Program Files\Git\bin\git.exe" config --global --add safe.directory D:\a\corral\corral
2026-03-06T00:42:35.1667193Z [command]"C:\Program Files\Git\bin\git.exe" config --local --name-only --get-regexp core\.sshCommand
2026-03-06T00:42:35.2033638Z [command]"C:\Program Files\Git\bin\git.exe" submodule foreach --recursive "sh -c \"git config --local --name-only --get-regexp 'core\.sshCommand' && git config --local --unset-all 'core.sshCommand' || :\""
2026-03-06T00:42:35.7516819Z [command]"C:\Program Files\Git\bin\git.exe" config --local --name-only --get-regexp http\.https\:\/\/github\.com\/\.extraheader
2026-03-06T00:42:35.7776396Z http.https://github.com/.extraheader
2026-03-06T00:42:35.7823296Z [command]"C:\Program Files\Git\bin\git.exe" config --local --unset-all http.https://github.com/.extraheader
2026-03-06T00:42:35.8126044Z [command]"C:\Program Files\Git\bin\git.exe" submodule foreach --recursive "sh -c \"git config --local --name-only --get-regexp 'http\.https\:\/\/github\.com\/\.extraheader' && git config --local --unset-all 'http.https://github.com/.extraheader' || :\""
2026-03-06T00:42:36.3979742Z Cleaning up orphan processes
2026-03-06T00:42:36.4299566Z Terminate orphan process: pid (5316) (vctip)

--- a/.release-notes/fix-heap-to-stack-crash.md
+++ b/.release-notes/fix-heap-to-stack-crash.md
@@ -1,0 +1,5 @@
+## Fix compiler crash in HeapToStack optimization pass on Windows
+
+The compiler could crash during optimization when compiling programs that trigger the HeapToStack pass (which converts heap allocations to stack allocations). The crash was non-deterministic and most commonly observed on Windows, though it could theoretically occur on any platform.
+
+The root cause was that HeapToStack was registered as a function-level optimization pass but performed operations (force-inlining constructors) that are only safe from a call-graph-level pass. This violated LLVM's pass pipeline contract and could cause use-after-free of internal compiler data structures.

--- a/src/libponyc/codegen/genopt.cc
+++ b/src/libponyc/codegen/genopt.cc
@@ -12,6 +12,8 @@
 
 #include <llvm/IR/PassManager.h>
 #include <llvm/Analysis/AliasAnalysis.h>
+#include <llvm/Analysis/CGSCCPassManager.h>
+#include <llvm/Analysis/LazyCallGraph.h>
 #include <llvm/Passes/OptimizationLevel.h>
 #include <llvm/Passes/PassBuilder.h>
 
@@ -86,26 +88,80 @@ public:
     c = compiler;
   }
 
-  PreservedAnalyses run(Function &f, FunctionAnalysisManager &am)
+  PreservedAnalyses run(LazyCallGraph::SCC &C, CGSCCAnalysisManager &AM,
+    LazyCallGraph &CG, CGSCCUpdateResult &UR)
   {
-    DominatorTree& dt = am.getResult<DominatorTreeAnalysis>(f);
-    BasicBlock& entry = f.getEntryBlock();
-    IRBuilder<> builder(&entry, entry.begin());
+    auto &FAM =
+      AM.getResult<FunctionAnalysisManagerCGSCCProxy>(C, CG).getManager();
 
     bool changed = false;
 
-    for(auto block = f.begin(), end = f.end(); block != end; ++block)
-    {
-      for(auto iter = block->begin(), end = block->end(); iter != end; )
-      {
-        Instruction* inst = &(*(iter++));
+    // Collect functions upfront to avoid iterator invalidation if
+    // updateCGAndAnalysisManagerForCGSCCPass splits the SCC.
+    SmallVector<Function*, 4> functions;
+    for(LazyCallGraph::Node &N : C)
+      functions.push_back(&N.getFunction());
 
-        if(runOnInstruction(builder, inst, dt, f))
-          changed = true;
-      }
+    for(Function *fp : functions)
+    {
+      if(fp->isDeclaration())
+        continue;
+
+      if(runOnFunction(*fp, FAM, CG, AM, UR))
+        changed = true;
     }
 
     return changed ? PreservedAnalyses::none() : PreservedAnalyses::all();
+  }
+
+  bool runOnFunction(Function &f, FunctionAnalysisManager &FAM,
+    LazyCallGraph &CG, CGSCCAnalysisManager &AM, CGSCCUpdateResult &UR)
+  {
+    bool changed = false;
+    bool restart;
+
+    do
+    {
+      restart = false;
+      DominatorTree& dt = FAM.getResult<DominatorTreeAnalysis>(f);
+      BasicBlock& entry = f.getEntryBlock();
+      IRBuilder<> builder(&entry, entry.begin());
+
+      for(auto block = f.begin(); block != f.end(); ++block)
+      {
+        for(auto iter = block->begin(); iter != block->end();)
+        {
+          Instruction* inst = &(*iter);
+
+          if(runOnInstruction(builder, inst, dt, f))
+          {
+            changed = restart = true;
+
+            FAM.invalidate(f, PreservedAnalyses::none());
+
+            auto *FN = CG.lookup(f);
+            if(FN)
+            {
+              auto *FC = CG.lookupSCC(*FN);
+              if(FC)
+              {
+                updateCGAndAnalysisManagerForCGSCCPass(
+                  CG, *FC, *FN, AM, UR, FAM);
+              }
+            }
+
+            break;
+          }
+
+          ++iter;
+        }
+
+        if(restart)
+          break;
+      }
+    } while(restart);
+
+    return changed;
   }
 
   bool runOnInstruction(IRBuilder<>& builder, Instruction* inst,
@@ -462,10 +518,6 @@ public:
     return false;
   }
 
-  void getAnalysisUsage(AnalysisUsage& use) const
-  {
-    use.addRequired<DominatorTreeWrapperPass>();
-  }
 };
 
 // Pass to replace pony_ctx calls in a dispatch function by the context passed
@@ -958,10 +1010,10 @@ static void optimise(compile_t* c, bool pony_specific)
   // Wire in the Pony-specific passes if requested.
   if(pony_specific)
   {
-    PB.registerPeepholeEPCallback(
-      [&](FunctionPassManager &fpm, OptimizationLevel level) {
+    PB.registerCGSCCOptimizerLateEPCallback(
+      [&](CGSCCPassManager &cgpm, OptimizationLevel level) {
         if(level.getSpeedupLevel() >= 2) {
-          fpm.addPass(HeapToStack(c));
+          cgpm.addPass(HeapToStack(c));
         }
       }
     );

